src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -110,54 +118,144 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
                 p != "startNode" && p != "endNode" && p != "space" && p != "externalId" && p != "type")
         ))
       .toSeq
+    val cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None
+    val select = Map("sync" -> SelectExpression(sources = sourceReference))
+    val syncMode =
+      decideSyncMode(cursors, instanceType, select).unsafeRunSync()
 
-    val instanceSyncRequest = InstanceSyncRequest(
-      `with` = Map("sync" -> tableExpression),
-      cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None,
-      select = Map("sync" -> SelectExpression(sources = sourceReference)),
-      includeTyping = Some(true)
-    )
-
-    Seq(syncOut(instanceSyncRequest, selectedColumns))
+    Seq(
+      syncOut(
+        syncMode,
+        `with` = Map("sync" -> tableExpression),
+        select,
+        selectedColumns
+      ))
   }
+  // scalastyle:on method.length
   // scalastyle:on cyclomatic.complexity
 
-  private def syncOut(syncRequest: InstanceSyncRequest, selectedProps: Array[String])(
-      implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] =
-    Stream.eval {
-      syncWithCursor(syncRequest).map { sr =>
-        val items = sr.items
-        val nextCursor = sr.nextCursor
-        val next =
-          (nextCursor, items.nonEmpty && items.last.lastUpdatedTime < terminationTimeStamp) match {
-            case (Some(cursor), true) =>
-              syncOut(syncRequest.copy(cursors = Some(Map("sync" -> cursor))), selectedProps)
-            case _ => fs2.Stream.empty
-          }
+  private val matchNothingFilter: FilterDefinition =
+    FilterDefinition.Not(MatchAll(JsonObject()))
 
-        val projected = items.map(toProjectedInstance(_, nextCursor, selectedProps))
-        fs2.Stream.emits(projected) ++ next
+  /**
+    * Validate cursors validity if present
+    * If cursors are invalid or not present generate sync cursor that would match _future_
+    * updates, demand initial full backfill fetch via query endpoint.
+    * Note: cursors validitiy is checked via future cursor fetch attempt so an attempt to fetch
+    *       future cursor is always made
+    */
+  private def decideSyncMode(
+      cursors: Option[Map[String, String]],
+      instanceType: InstanceType,
+      select: Map[String, SelectExpression]): IO[SyncMode] =
+    fetchData(
+      useQueryEndpoint = false,
+      cursors = cursors,
+      `with` = Map("sync" -> generateTableExpression(instanceType, matchNothingFilter)),
+      select = select)
+      .flatMap { sr =>
+        (sr.nextCursor, cursors) match {
+          case (None, _) =>
+            IO.raiseError(
+              new IllegalArgumentException("/sync response must have " +
+                "nextCursor"))
```

Comment:
https://github.com/cognitedata/cognite-sdk-scala/pull/730 to make /sync nextCursor non-required